sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_launcher.py

```diff
@@ -0,0 +1,199 @@
+from apache_beam.options.pipeline_options import StandardOptions
+from apache_beam.options.pipeline_options import TestOptions
+
+from nexmark_util import Command
+from queries import query0
+
+logging.basicConfig(level=logging.INFO,
+                    format='(%(threadName)-10s) %(message)s')
+
+# Reserved configuration to run these benchmark suite.
+PUBSUB_PROJECT = 'google.com:clouddfe'
+PUBSUB_SUBSCRIPTION = 'nexmark-benchmark'
+PUBSUB_TOPIC = 'nexmark-benchmark'
+
+# TODO(mariagh): Make comprehensive and large
+EVENT_COUNT_PER_TYPE = 3
+PERSON_EVENT = 'p12345,maria,maria@maria.com,1234-5678-9012-3456,sunnyvale,CA,1528098831536' # pylint: disable=line-too-long
+AUCTION_EVENT = 'a12345,car,2012 hyundai elantra,20K,15K,1528098831536,20180630,maria,vehicle' # pylint: disable=line-too-long
+BID_EVENT = 'b12345,maria,354,1528098831536'
+
+
+def parse_args():
+  parser = argparse.ArgumentParser()
+
+  parser.add_argument('--query', '-q',
+                      type=int,
+                      action='append',
+                      required=True,
+                      choices=[0, 1],
+                      help='Query to run')
+
+  parser.add_argument('--subscription',
+                      type=str,
+                      help='Pub/Sub subscription to read from')
+
+  parser.add_argument('--topic',
+                      type=str,
+                      help='Pub/Sub topic to read from')
+
+  args, pipeline_args = parser.parse_known_args()
+  options = PipelineOptions(pipeline_args)
+  print('args, pipeline_args:', args, pipeline_args)
+
+  # Usage with Dataflow requires a project to be supplied.
+  if options.view_as(GoogleCloudOptions).project is None:
+    parser.print_usage()
+    print(sys.argv[0] + ': error: argument --project is required')
+    sys.exit(1)
+
+  # Pub/Sub is currently available for use only in streaming pipelines.
+  if options.view_as(StandardOptions).streaming is None:
+    parser.print_usage()
+    print(sys.argv[0] + ': error: argument --streaming is required')
+    sys.exit(1)
+
+  # wait_until_finish ensures that the streaming job is canceled.
+  if options.view_as(TestOptions).wait_until_finish_duration is None:
+    parser.print_usage()
+    print(sys.argv[0] + ': error: argument --wait_until_finish_duration is required') # pylint: disable=line-too-long
+    sys.exit(1)
+
+  # We use the save_main_session option because one or more DoFn's in this
+  # workflow rely on global context (e.g., a module imported at module level).
+  options.view_as(SetupOptions).save_main_session = True
+  return args, options
+
+
+def generate_events(pipeline, args, pipeline_options):
+  from google.cloud import pubsub
+  publish_client = pubsub.Client(project=PUBSUB_PROJECT)
+  topic = publish_client.topic(PUBSUB_TOPIC)
+  sub = topic.subscription(PUBSUB_SUBSCRIPTION)
+  assert topic.exists()
+  assert sub.exists()
+
+  logging.info('Generating %d auction events to topic %s',
```

Comment:
This is a strange model. It generates all events ahead of time, before starting the job. This might be fine for this level of testing. Have you considered running this in a thread in parallel?

Reply:
Yes, I have. I was just trying to avoid having a situation where the generator of events could become the bottleneck, since what we want is performance benchmarks.